README.md
## Simple API
A simple web API built with Ruby 3 and Sinatra Framework.
Running in Heroku.

[![<joesustaric>](https://circleci.com/gh/joesustaric/simple-api.svg?style=svg)](https://app.circleci.com/pipelines/github/yjoesustaric/simple-api?branch=main)

## API

`/`

Returns `Hello World`.

`/healthcheck`

Returns `200` if the API service is healthy.

Example JSON response.
```json
{
    "status": "unhealthy"
}
```

`/metadata`

Returns `200` if the API service is healthy.
Returns `503` if service is unhealthy.

Example JSON response.
```json
{
    "myapplication":[
        {
            "version": "1.3.0",
            "description": "simple api",
            "lastcommitsha": "c4c1f203401d9491e07d6c5ab771e5d0eed6dc36"
        }
    ]
}
```

## Development

This repository has a `.tools-version` file. You can set up [asdf](http://asdf-vm.com/guide/getting-started.html#_1-install-dependencies) to manage the versions of dependencies listed in that file. Currently it is only Ruby 3.0.0. Once you install `asdf` you will need to install the Ruby 3.0.0 version.

If you have your own version manager or ruby 3.0.0 alredy installed then you can skip this step.

After checking out the repo, run `bin/setup` to install dependencies. Then, run `rake spec` to run the tests. You can also run `bin/console` for an interactive prompt that will allow you to experiment.

To run the tests and the code linter, run:
```
bundle exec rake
```

To run the app locally after all the dependencies are installed the easiest way is by:
```
APP_ENV=development exe/simple-api
```

The default site and port is `http://127.0.0.1:4567` but you can override the port number by specifying a `PORT` eg `PORT=8080 APP_ENV=development exe/simple-api`.

## Deployment
This `simple-api` will be auto deployed to Heroku on every main green build via the CircleCI [test and deploy pipeline](https://circleci.com/gh/joesustaric/simple-api).
You may need to sign into CircleCi with your GitHub account to see the build.

It is packaged up in a Docker container and uses the [Heroku Container Registry](https://devcenter.heroku.com/articles/container-registry-and-runtime) method of deploying.
The production site is `https://joes-simple-api.herokuapp.com/`.

## Monitoring
If you have access to the Heroku Application, you can simply install the Heroku CLI via [this guide](https://devcenter.heroku.com/articles/heroku-cli).

Then to inspect the running application logs run:
```
heroku logs --tail -a joes-simple-api
```

## Extensions
Not in priority order.

- Add in ["rack/protection"](http://sinatrarb.com/rack-protection/) for protection from typical web attacks.
- Add in [dependabot](https://dependabot.com/) for code vulnrability scanning.
- Container scanning.
- Use a different API key with restricted permissions to deploy to Heroku.
- Adding a job that tests the latest Ruby version compatabiliity.
- PR merge blocks on failed builds.
- Test coverage gate. Fail build on x% or below test coverage.
- Refactor code into controllers.
- More monitoring and observability.

## Contributing
PR's Welcome.

## License
[MIT](LICENSE)


CHANGELOG.md
## [Unreleased]

## [1.3.2] - 2021-08-25

### Fixed
- Writing the REVISION file.

## [1.3.1] - 2021-08-25

### Fixed
- `bundle install --without test development` for production

## [1.3.0] - 2021-08-25

### Added
- GET `/metadata` now returns the latest code commit sha

## [1.2.0] - 2021-08-24

### Added
- GET `/metadata`

## [1.1.2] - 2021-08-24

### Added
- json response for healthcheck

## [1.1.1] - 2021-08-24

### Added
- Remove Rack protection not needed.

## [1.1.0] - 2021-08-24

### Added
- GET `/helthcheck`
- Rack protection

## [1.0.0] - 2021-08-23

### Added
- GET `/` = Hello World

## [0.1.0] - 2021-08-22

- Initial release


spec/spec_helper.rb
# frozen_string_literal: true

$LOAD_PATH << File.expand_path("../lib/simple_api", __dir__)

require "rspec"
require "rack/test"
require "simple_api"

ENV["APP_ENV"] = "test"

RSpec.configure do |config|
  # Enable flags like --only-failures and --next-failure
  config.example_status_persistence_file_path = ".rspec_status"

  config.include Rack::Test::Methods

  # Disable RSpec exposing methods globally on `Module` and `main`
  config.disable_monkey_patching!

  config.expect_with :rspec do |c|
    c.syntax = :expect
  end
end


spec/simple/simple_api_spec.rb
# frozen_string_literal: true

require "json"
# rubocop:disable Metrics/BlockLength
RSpec.describe "The SimpleApi App" do
  def app
    SimpleApi::Server
  end

  it "has a version number" do
    expect(SimpleApi::VERSION).not_to be nil
  end

  describe "/" do
    it "says hello" do
      get "/"

      expect(last_response).to be_ok
      expect(last_response.body).to eq("Hello World\n")
    end
  end

  describe "/healthcheck" do
    it "responds with 200 when healthy" do
      get "/healthcheck"

      expect(last_response).to be_ok
      expect(last_response.body).to eq({ status: "healthy" }.to_json)
    end
  end

  describe "/metadata" do
    it "responds with json metadata" do
      response = {
        myapplication: [
          {
            version: SimpleApi::VERSION,
            description: "simple api",
            lastcommitsha: "abc57858585"
          }
        ]
      }.to_json

      get "/metadata"

      expect(last_response).to be_ok
      expect(last_response.body).to eq response
    end
  end
end
# rubocop:enable Metrics/BlockLength


spec/simple_api/simple_api_spec.rb
# frozen_string_literal: true

require "json"
# rubocop:disable Metrics/BlockLength
RSpec.describe "The SimpleApi App" do
  def app
    SimpleApi::Server
  end

  it "has a version number" do
    expect(SimpleApi::VERSION).not_to be nil
  end

  describe "/" do
    it "says hello" do
      get "/"

      expect(last_response).to be_ok
      expect(last_response.body).to eq("Hello World\n")
    end
  end

  describe "/healthcheck" do
    context "Given there is a REVISION file with content" do
      before { File.open("REVISION", "w") { |f| f.write "a git sha\n" } }

      context "When we call /healthcheck" do
        it "responds with 200 when healthy" do
          get "/healthcheck"

          expect(last_response.status).to eq 200
          expect(last_response.body).to eq({ status: "healthy" }.to_json)
        end
      end

      after { File.delete("REVISION") if File.exist?("REVISION") }
    end

    context "Given there is no REVISION file" do
      before { File.delete("REVISION") if File.exist?("REVISION") }

      context "When we call /healthcheck" do
        it "responds with 503 (service unavailable) " do
          get "/healthcheck"

          expect(last_response.status).to eq 503
          expect(last_response.body).to eq({ status: "unhealthy" }.to_json)
        end
      end
    end
  end

  describe "/metadata" do
    context "Given the service is healthy" do
      before do
        File.open("REVISION", "w") { |f| f.write "some-git-sha\n" }
      end
      context "When we call /metadata" do
        it "responds with json metadata" do
          response = {
            myapplication: [
              {
                version: SimpleApi::VERSION,
                description: "simple api",
                lastcommitsha: "some-git-sha"
              }
            ]
          }.to_json

          get "/metadata"

          expect(last_response).to be_ok
          expect(last_response.body).to eq response
        end

        after do
          File.delete("REVISION") if File.exist?("REVISION")
        end
      end
    end

    context "Given the service is unhealthy" do
      before { File.delete("REVISION") if File.exist?("REVISION") }

      context "When we call /metadata" do
        it "responds with json metadata" do
          response = {
            myapplication: [
              {
                version: SimpleApi::VERSION,
                description: "simple api",
                lastcommitsha: "no commit sha found"
              }
            ]
          }.to_json

          get "/metadata"

          expect(last_response).to be_ok
          expect(last_response.body).to eq response
        end

        after do
          File.delete("REVISION") if File.exist?("REVISION")
        end
      end
    end
  end
end
# rubocop:enable Metrics/BlockLength


lib/simple_api/server.rb
# frozen_string_literal: true

require "sinatra"
require "json"

# https://ruby-doc.org/core-3.0.0/IO.html#method-i-sync-3D
$stdout.sync = true
$stderr.sync = true

module SimpleApi
  class Error < StandardError; end

  # The Server API
  class Server < Sinatra::Application
    get "/" do
      "Hello World\n"
    end

    get "/healthcheck" do
      content_type :json

      health = "unhealthy"
      code = 503

      if File.exist?("REVISION")
        health = "healthy"
        code = 200
      end

      body({ status: health.to_s }.to_json)
      status code
    end

    get "/metadata" do
      content_type :json

      last_commit_sha = "no commit sha found"
      last_commit_sha = IO.read("REVISION") if File.exist?("REVISION")

      {
        myapplication: [
          {
            version: VERSION,
            description: "simple api",
            lastcommitsha: last_commit_sha.strip.to_s
          }
        ]
      }.to_json
    end
  end
end
